nat.go

```diff
@@ -141,36 +112,58 @@ func (nat *NAT) NewMapping(protocol string, port int) (Mapping, error) {
 	}
 
 	m := &mapping{
-		intport: port,
-		nat:     nat,
-		proto:   protocol,
+		intport:  port,
+		nat:      nat,
+		proto:    protocol,
+		teardown: nat.removeMapping,
 	}
 
-	m.proc = goprocess.WithTeardown(func() error {
-		nat.rmMapping(m)
-		nat.natmu.Lock()
-		defer nat.natmu.Unlock()
-		nat.nat.DeletePortMapping(m.Protocol(), m.InternalPort())
-		return nil
-	})
-
-	nat.addMapping(m)
-
-	m.proc.AddChild(periodic.Every(MappingDuration/3, func(worker goprocess.Process) {
-		nat.establishMapping(m)
-	}))
+	nat.mappingmu.Lock()
+	if nat.closed {
+		nat.mappingmu.Unlock()
+		return nil, errors.New("closed")
+	}
+	nat.mappings[m] = struct{}{}
+	nat.refCount.Add(1)
+	nat.mappingmu.Unlock()
+	go nat.refreshMappings(m)
```

Comment:
If it's not too difficult, it would be nice to have a single worker loop to handle refreshing mappings.

Reply:
Yes, in a follow-up PR.

Reply:
Unfortunately, it's not that easy, because of the way that `mapping` and `nat` are coupled. Not sure I want to make such a change in a repo where we don't have any tests...